Application/EdFi.Ods.Api/Configuration/AutoEncryptingOdsInstanceConfigurationDataProviderDecorator.cs

```diff
@@ -0,0 +1,85 @@
+    private readonly IEdFiAdminRawOdsInstanceConfigurationDataProvider _edFiAdminRawOdsInstanceConfigurationDataProvider;
+    private readonly Lazy<byte[]> _privateKeyBytesLazy;
+
+    public AutoEncryptingOdsInstanceConfigurationDataProviderDecorator(
+        ISymmetricStringEncryptionProvider symmetricStringEncryptionProvider,
+        ISymmetricStringDecryptionProvider symmetricStringDecryptionProvider,
+        IEdFiOdsConnectionStringWriter edFiOdsConnectionStringWriter, 
+        IAdminDatabaseConnectionStringProvider adminDatabaseConnectionStringProvider,
+        DbProviderFactory dbProviderFactory,
+        Lazy<byte[]> privateKeyBytesLazy)
+    {
+        _symmetricStringEncryptionProvider = symmetricStringEncryptionProvider;
+        _symmetricStringDecryptionProvider = symmetricStringDecryptionProvider;
+        _edFiOdsConnectionStringWriter = edFiOdsConnectionStringWriter;
+        _edFiAdminRawOdsInstanceConfigurationDataProvider = new EdFiAdminRawOdsInstanceConfigurationDataProvider(adminDatabaseConnectionStringProvider, dbProviderFactory);
+        _privateKeyBytesLazy = privateKeyBytesLazy;
+    }
+
+    /// <inheritdoc cref="IEdFiAdminRawOdsInstanceConfigurationDataProvider.GetByIdAsync" />
+    public async Task<RawOdsInstanceConfigurationDataRow[]> GetByIdAsync(int odsInstanceId)
+    {
+        var rawDataRows = await _edFiAdminRawOdsInstanceConfigurationDataProvider.GetByIdAsync(odsInstanceId);
+
+        if (!System.Diagnostics.Debugger.IsAttached)
+        {
+            foreach (var row in rawDataRows)
+            {
+                row.ConnectionString = ApplyConnectionStringEncryptionLogic(row);
+            }
+        }
+
+        return rawDataRows;
+    }
+
+    /// <summary>
+    /// Encrypts the connection string in the database if it is not already encrypted, and returns the plaintext connection string.
+    /// </summary>
+    /// <param name="rawDataRow">
+    /// The <see cref="RawOdsInstanceConfigurationDataRow"/> containing the connection string to which the encryption logic should be applied.
+    /// </param>
+    /// <returns>
+    /// Plaintext connection string.
+    /// </returns>
+    private string ApplyConnectionStringEncryptionLogic(RawOdsInstanceConfigurationDataRow rawDataRow)
+    {
+        var privateKey = _privateKeyBytesLazy.Value;
+        
+        if (_symmetricStringDecryptionProvider.TryDecrypt(
+                rawDataRow.ConnectionString, out string decryptedConnectionString, privateKey))
+        {
+            return decryptedConnectionString;
+        }
+
+        // If decryption fails, then we assume the connection string is not encrypted, so we encrypt it and write it to the database.
+        var plaintextConnectionString = rawDataRow.ConnectionString;
+        var encryptedConnectionString = _symmetricStringEncryptionProvider.Encrypt(rawDataRow.ConnectionString, privateKey);
+
+        _edFiOdsConnectionStringWriter.WriteConnectionString(
```

Comment:
With the bug above, this call (which passes an actual derivative type when it's present) would have the effect of overwriting the derivative type connection string with the encrypted version of the main ODS connection string (except that the query to update the derivative connection string throws a runtime exception -- see below.)